### PR TITLE
refactor formMethods and control object (use more js and less react for performance reasons)

### PR DESCRIPTION
**Pull request by mh-alahdadian**
split one of custom hook logics and use pure-js for better performance and better control of breaking logics down

next changes will be split all logics and move core logics to a single file and other hooks could be wrappers around it 
and also writing less `.current` for refs will decrease bundle size 
even it could be possible to completely break logic into separate repo named `@hookform/core` and some hook utilities around core named `@hookform/react`

also in future maybe we could export `createForm` function to be used out of react components

### CODE REVIEW

**Comment by MHA15** [comment]
in third commit there is a braking change in using control object
control object doesn't have many ref objects inside itself instead we can read latest value by getting it from control 
it means these are equal `control.x.current` is now `control.x` and also for accessing last value you should pass control itself and we can not pass `control.x` as a ref object anymore

**Comment by MHA15** [comment]
I just realized that one of the most important breaking changes of this PR is that useForm methods is equal per renders and this is a huge performance improvement

**Comment by MHA15** [comment]
@bluebill1049 
I would be happy if you review this PR and let me know

**Comment by MHA15** [comment]
It could be a good idea to remove control setters which are not used/should not used , but I'm not sure about them
And also in future whole core logics could be integrated in createForm and can change whole control type (mostly have internal methods instead of exporting private values)

**Comment by bluebill1049** [comment]
I will get back to you.

**Comment by MHA15** [comment]
any update ?

**Comment by bluebill1049** [comment]
still got a few more PR ahead of this one, will get to it.

**Comment by MHA15** [comment]
> still got a few more PR ahead of this one, will get to it.

I think that would be a good idea to read PR description and feedback to it before really reviewing code of it

**Comment by bluebill1049** [comment]
Thanks very much for the amazing work! I finally get time to look into it a bit more this time. I would recommend having a discussion next time before a big change like this.

**Comment by MHA15** [comment]
Yeah I stop continuing branch to have discussion first, and as I said third commit has all braking changes and if we don't want them we could merge only first two commits

**Comment by bluebill1049** [comment]
Definitely some interesting and great ideas here! Once this [PR](https://github.com/react-hook-form/react-hook-form/pull/5958) is solid and merged and I will come back to this. Maybe it's worth doing some performance comparison.

>  it could be possible to completely break logic into separate repo named @hookform/core and some hook utilities around core named @hookform/react

Love this idea above as well.

**Comment by bluebill1049** [comment]
> all braking changes

yea, probably avoid breaking changes for now.

File: src/useForm.ts [hunk 1]
```diff
@@ -82,21 +82,23 @@ import {
...
 const isWindowUndefined = typeof window === 'undefined';
 
-export function useForm<
+const defaultOptions = {
+  mode: VALIDATION_MODE.onSubmit,
+  reValidateMode: VALIDATION_MODE.onChange,
+  shouldFocusError: true,
+} as const;
+
+export function createForm<
```
Comment (bluebill1049) [inline src/useForm.ts:91]: I really like this approach.

**Comment by bluebill1049** [comment]
Do you mind rever back to the second commit? or else I can do that later as well.

**Comment by bluebill1049** [comment]
never mind, i will sort it out later 👍

**Comment by MHA15** [comment]
But this control refactor is a good approach and can help us to receive the vision I told in first commit 
Is there any problem with this breaking change ? let's edit other usages too. 
Or  do you want to let this refactor to merge right before V8  ?

**Comment by bluebill1049** [comment]
Don't worry i will sort it out.

**Comment by bluebill1049** [comment]
If you want to review this PR: https://github.com/react-hook-form/react-hook-form/pull/5987

**Comment by MHA15** [comment]
should we close this merge request ?

**Comment by bluebill1049** [comment]
> should we close this merge request ?

yes, plz

**Pull request closed without merging**